src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,411 @@
+import org.apache.lucene.search.FieldDoc;
+import org.apache.lucene.util.PriorityQueue;
+import org.opensearch.neuralsearch.query.HybridQueryScorer;
+import org.opensearch.common.Nullable;
+import org.opensearch.neuralsearch.search.HitsThresholdChecker;
+import org.opensearch.neuralsearch.search.MultiLeafFieldComparator;
+
+/*
+ Collects the TopFieldDocs after executing hybrid query. Uses HybridQueryTopDocs as DTO to handle each sub query results.
+ The individual query results are sorted as per the sort criteria sent in the search request.
+ */
+@Log4j2
+public abstract class HybridTopFieldDocSortCollector implements Collector {
+    /*
+      numhits maintain the size of the result to be collected.
+     */
+    final int numHits;
+    /*
+      hitsThresholdChecker is used to get score mode and check the threshold for collecting the hits.
+     */
+    final HitsThresholdChecker hitsThresholdChecker;
+    /*
+      docBase holds the starting point of doc Iteration in the segment.
+     */
+    int docBase;
+    /*
+      comparators collect the value of the field on which sorting criteria is applied and returns the result accordingly.
+     */
+    LeafFieldComparator comparators[];
+    /*
+      reverseMul is used to set the direction of the sorting when creating comparators.
+      In threshold check reverseMul is used in comparison logic.
+      It modifies the comparison of either reverse or maintain the natural order depending on its value.
+      This ensures that the compareBottom method adjusts the order based on whether you want ascending or descending sorting.
+     */
+    int reverseMul;
+    FieldComparator<?> firstComparator;
+    FieldValueHitQueue.Entry bottom = null;
+    /*
+      List of Priority Queues to hold entries which has higher score but sorted as per the sorting criteria.
+     */
+    FieldValueHitQueue<FieldValueHitQueue.Entry>[] compoundScores;
+    boolean queueFull[];
+    @Getter
+    private int totalHits;
+    @Getter
+    float maxScore = 0.0f;
+    int[] collectedHits;
+    int numberOfSubQueries = 0;
+    @Getter
+    @Setter
+    private TotalHits.Relation totalHitsRelation = TotalHits.Relation.EQUAL_TO;
+    /*
+       searchSortPartOfIndexSort is used to evaluate whether to perform index sort or not.
+     */
+    private Boolean searchSortPartOfIndexSort = null;
+
+    private static final TopFieldDocs EMPTY_TOPDOCS = new TopFieldDocs(
+        new TotalHits(0, TotalHits.Relation.EQUAL_TO),
+        new ScoreDoc[0],
+        new SortField[0]
+    );
+
+    // Declaring the constructor private prevents extending this class by anyone
+    // else. Note that the class cannot be final since it's extended by the
+    // internal versions. If someone will define a constructor with any other
+    // visibility, then anyone will be able to extend the class, which is not what
+    // we want.
+    public HybridTopFieldDocSortCollector(final int numHits, final HitsThresholdChecker hitsThresholdChecker) {
```

Comment:
since this is an abstract class lets make the constructor package private so that classes in this package only can use this constructor.